### ISSUE TITLE
No episode was completed since last episode

### ISSUE THREAD

**Issue by roboserg** [Unity-Technologies/ml-agents#2200]
So I am doing a simple flappy bird bot - https://puu.sh/DLHI2/b7060a2db2.png

An episode ends with Done() if the agent hits either the ground, ceiling, or the towers. Max steps in both the agent and the academy are set to 0. 

**1.** 
Yet for some reason after several hundred of thousands of training steps I start getting the "No episode was completed since last episode" - https://puu.sh/DLHFX/70c2074a80.png

Why does it happen and how do I fix it?

**2.** 
I am also unsure why the cumulative reward jumps like that - https://puu.sh/DLHMh/cd34aa1b6d.png
I tried gamma 0.9 / 0.95 / 0.999 and also different batch_size with 10 / 100 / 1000 (buffer_size was always 10x of the batch_size). Any tips or suggestions of why it jumps like that and how to improve the learning?  The variance is too high, the agent does only 7-30 towers on average. It should fly almost indefinitely

ps. rewards are -1 when hitting an obstacle. +1 when flying through the tower and +0.001 at every step alive. Two discreet actions, 0 do nothing and 1 jump.

pss. training config:

```
default:
    max_steps: 1000000
    batch_size: 1000
    buffer_size: 10000
    gamma: 0.9
    epsilon: 0.2
    trainer: ppo
    beta: 5.0e-3
    hidden_units: 128
    lambd: 0.95
    learning_rate: 3.0e-4
    memory_size: 256
    normalize: true
    num_epoch: 3
    num_layers: 2
    time_horizon: 64
    sequence_length: 64
    summary_freq: 500
    use_recurrent: false
    use_curiosity: false
    curiosity_strength: 0.01
    curiosity_enc_size: 128
```

**Comment by djsdvg** [comment]
"No episode was completed since last episode" means the after ``summary_freq`` (500 in your case) steps the episode is not done yet.
So it's a good thing because it indicates that the agent (the bird)  is learning and is able to fly longer without hit the pipes.
In fact if you notice, after a few "No episode" message the reward is higher.

**Comment by roboserg** [comment]
Thanks! Whew, so "no episode" is actually good and it all makes sense now ;)

Here are the curves you requested - https://puu.sh/DLVWH/c504f8053d.png
The variance in the reward curve is astronomical, and I dont understand why - https://puu.sh/DLVXZ/6454dee1ff.png

**Comment by djsdvg** [comment]
It seems that you need more steps.
You can also use ``--num-envs``to train in parallel and get better results faster.
Also check the observation that you give to your agent, maybe there's some hidden state.

**Comment by roboserg** [comment]
@djsdvg 
If it needs more steps to train that would be exactly the issue I can not figure out. As I mentioned earlier the rocket environment with 3 continuous actions and 6 observations needed less then 1 mio. steps to train. And this flappy bird after 1 mio. steps has a huge variance in the reward.

I am testing a new run now with 4 observations. Before it had vertical velocity as well as relative vertical and horizontal distance from the agent to the mid point of the opening between two pipes. Now added the absolute vertical height as observation. Also now training in headless mode with 16x instances. The variance is still too high and seems very similar https://puu.sh/DM2E1/f240c15b50.png (green new with 4 observations and 16x instances, red old training run)

**Comment by djsdvg** [comment]
Have you setted a max steps limit for the agents?
If not it could be the cause of your problem.

(If you don't the set a limit in steps there's no limit in reward too.
So the agent will continue to beat himself and make new records in game.
If there's no a max the entropy will stay high and that will take to unstable rewards.) <--I'm not 100% sure about that

**Comment by roboserg** [comment]
So I tested with max steps = 5000, the agent could reach 25 pipes with 0 variance. Then I changed so that the episode ends not only after death but also after 250 pipes. So 250 pipes is the maximum. With 250 pipes I see very similar, high variance training run as before - https://puu.sh/DMsle/365777315c.png
250 pipes run with:
```
batch_size: 128
buffer_size: 4048
gamma: 0.99
```

**Comment by roboserg** [comment]
Did another run with 14 instances - https://puu.sh/DNa8z/8a766ac9bf.png

Why would the cum. reward drop and loss start to rise after ca. 150k steps? I find the training behavior very weird.
```
batch_size: 200
buffer_size: 10000
gamma: 0.99
```

**Comment by roboserg** [comment]
OK I solved the game at last, the agent does > 2000 pipes and practically never dies (0 variance for 250 pipes case scenario) - https://puu.sh/DNj71/9db8dedb61.png

https://puu.sh/DNj2y/18e639592e.mp4

I think the problem was in the observation, I was feeding it "0" distance to the pipes if none were visible (at the beginning of the game). Feeding "-1" instead seemed to helped (along with other small fixes).

3h of training / 700k steps with 18 environments. Does that mean the whole training took 700.000 * 18 = 12.6 million steps?
